frame/reward-curve/src/lib.rs

```diff
@@ -0,0 +1,125 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Kulupu is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Kulupu. If not, see <http://www.gnu.org/licenses/>.
+
+//! Set the block reward with a reward curve.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::{Encode, Decode};
+use sp_std::prelude::*;
+use sp_runtime::traits::Zero;
+use frame_support::{decl_storage, decl_module, decl_error, decl_event, ensure, weights::Weight};
+use frame_support::traits::{Currency, LockableCurrency, Get, EnsureOrigin};
+use pallet_rewards::SetReward;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+pub trait Trait: frame_system::Trait {
+	/// The overarching event type.
+	type Event: From<Event> + Into<<Self as frame_system::Trait>::Event>;
+	/// An implementation of on-chain currency.
+	type Currency: LockableCurrency<Self::AccountId>;
+	/// How often to check to update the reward curve.
+	type UpdateFrequency: Get<Self::BlockNumber>;
+	/// The origin that can set the reward curve.
+	type UpdateOrigin: EnsureOrigin<Self::Origin>;
+	/// Handler for setting a new reward.
+	type SetReward: SetReward<BalanceOf<Self>>;
+}
+
+/// Type alias for currency balance.
+pub type BalanceOf<T> = <<T as Trait>::Currency as Currency<<T as frame_system::Trait>::AccountId>>::Balance;
+
+#[derive(Encode, Decode, Clone, PartialEq, Debug)]
+pub struct RewardPoint<BlockNumber, Balance> {
+	start: BlockNumber,
+	reward: Balance,
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		/// Reward curve is not sorted.
+		NotSorted,
+	}
+}
+
+decl_event! {
+	pub enum Event {
+		/// A new reward curve was set.
+		RewardCurveSet,
+		/// Reward updated successfully.
+		UpdateSuccessful,
+		/// Reward failed to update.
+		UpdateFailed,
+	}
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Eras {
+		/// Reward Curve for this chain
+		pub RewardCurve get(fn reward_curve) config(): Vec<RewardPoint<T::BlockNumber, BalanceOf<T>>>;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event() = default;
+
+		fn on_initialize(current_block: T::BlockNumber) -> Weight {
+			let mut weight: Weight = 0;
+			if current_block % T::UpdateFrequency::get() == Zero::zero() {
+				let _ = RewardCurve::<T>::try_mutate(|curve| -> Result<(), ()> {
+					weight = weight.saturating_add(T::DbWeight::get().reads(1));
+					ensure!(!curve.is_empty(), ());
+					// We checked above that curve is not empty, so this will never panic.
+					let point = curve.remove(0);
+					ensure!(point.start <= current_block, ());
+					let new_reward = point.reward;
+					// Not much we can do if this fails.
```

Comment:
Thinking about this @shawntabrizi, I'm right now actually slightly leaning towards putting this functionality into pallet-rewards. If we have higher degree of integration, we can directly check the validity of the rewards in `set_reward_curve` and avoid the ugliness here of ignoring `set_reward` failures. WDYT?

Reply:
For the point of this being an optional feature -- `pallet-rewards` is already entirely Kulupu-specific. The "removability" you mentioned (for having a separate module) is also less of a concern IMO. When we are done with the curve, even in a single module we can still easily remove it. Dispatchable index may change, but for that we just need to bump the transaction version. Anything that requires a migration still requires a migration.

Reply:
I am happy enough to refactor to however you see fit. Not that the only way the update fails is if it does not pass the "at least bigger than existential deposit" check.

If this failure is a concern, we could program the `set_reward` function to directly adjust the storage item without checks, thus there will never be a fail condition.

But, to me it is all the same.

Reply:
> Note that the only way the update fails is if it does not pass the "at least bigger than existential deposit" check.

Yes. However, we can check the rewards first in `set_reward_curve`. Then when we actually set the reward, we can be assured that it's always "at least bigger than existential deposit". That way it's guaranteed to always pass.

However, this requires a tighter integration with pallet-rewards. If we do a separate pallet, we need to have an explicit dependency on pallet-rewards and also mark the `Reward` storage item as `pub` (which is currently not the case). This is why I'm asking whether it makes sense to put this functionality directly in pallet-rewards.